### PR TITLE
Refactor/Refactor and document `cid`, `oid` package functionality

### PR DESCRIPTION
**Pull request by carpawell**

### CODE REVIEW

File: audit/result.go [hunk 1]
```diff
@@ -76,16 +77,28 @@ func (r *Result) ForEpoch(epoch uint64) {
 // Return value MUST NOT be mutated: to do this, first make a copy.
 //
 // See also ForContainer.
-func (r Result) Container() *cid.ID {
-	return cid.NewFromV2(r.v2.GetContainerID())
+func (r Result) Container() cid.ID {
+	var cID cid.ID
+
+	cidV2 := r.v2.GetContainerID()
+	if cidV2 == nil {
```
Comment (cthulhu-rider) [inline audit/result.go:84]: Can be shortened:
```
if cidV2 != nil {
  _ = cID.ReadFromV2(*cidV2)
}
```

File: audit/result.go [hunk 2]
```diff
@@ -76,16 +77,28 @@ func (r *Result) ForEpoch(epoch uint64) {
...
+		return cID
+	}
+
+	_ = cID.ReadFromV2(*cidV2)
```
Comment (cthulhu-rider) [inline audit/result.go:88]: When `error` can occur?

File: audit/result.go [hunk 3]
```diff
@@ -76,16 +77,28 @@ func (r *Result) ForEpoch(epoch uint64) {
 // Return value MUST NOT be mutated: to do this, first make a copy.
 //
 // See also ForContainer.
-func (r Result) Container() *cid.ID {
-	return cid.NewFromV2(r.v2.GetContainerID())
+func (r Result) Container() cid.ID {
```
Comment (cthulhu-rider) [inline audit/result.go:80]: If return type is changed, then docs must be fixed.

File: audit/result.go [hunk 4]
```diff
@@ -76,16 +77,28 @@ func (r *Result) ForEpoch(epoch uint64) {
...
+	if cidV2 == nil {
+		return cID
+	}
+
+	_ = cID.ReadFromV2(*cidV2)
+
+	return cID
 }
 
 // ForContainer returns identifier of the container with which the data audit Result
```
Comment (cthulhu-rider) [inline audit/result.go:93]:
```suggestion
// ForContainer sets identifier of the container with which the data audit Result
```

File: client/object_get.go [hunk 5]
```diff
@@ -35,10 +35,10 @@ type prmObjectRead struct {
 	bearer    token.BearerToken
 
 	cnrSet bool
-	cnr    cid.ID
+	cnrID  cid.ID
```
Comment (cthulhu-rider) [inline client/object_get.go:38]: Why?

File: container/announcement.go [hunk 6]
```diff
@@ -38,15 +39,25 @@ func (a *UsedSpaceAnnouncement) SetEpoch(epoch uint64) {
...
+
+	_ = cID.ReadFromV2(*cidV2)
+
+	return
 }
 
 // SetContainerID sets announcement container value.
-func (a *UsedSpaceAnnouncement) SetContainerID(cid *cid.ID) {
-	(*container.UsedSpaceAnnouncement)(a).SetContainerID(cid.ToV2())
+func (a *UsedSpaceAnnouncement) SetContainerID(cid cid.ID) {
```
Comment (cthulhu-rider) [inline container/announcement.go:56]: Package name collision
```suggestion
func (a *UsedSpaceAnnouncement) SetContainerID(cnr cid.ID) {
```

File: container/id/id.go [hunk 7]
```diff
@@ -1,90 +1,112 @@
...
+// 	_ = ID([32]byte) // not recommended
+type ID [sha256.Size]byte
 
-// SetSHA256 sets container identifier value to SHA256 checksum of container body.
-func (id *ID) SetSHA256(v [sha256.Size]byte) {
-	(*refs.ContainerID)(id).SetValue(v[:])
+// ReadFromV2 reads ID from the refs.ContainerID message.
+//
+// See also WriteToV2.
+func (id *ID) ReadFromV2(m refs.ContainerID) error {
```
Comment (cthulhu-rider) [inline container/id/id.go:25]: Lets mention `error` return in docs.

File: container/id/id.go [hunk 8]
```diff
@@ -1,90 +1,112 @@
...
-func (id *ID) Equal(id2 *ID) bool {
-	return bytes.Equal(
-		(*refs.ContainerID)(id).GetValue(),
-		(*refs.ContainerID)(id2).GetValue(),
-	)
+// Encode encodes ID into 32 bytes of dst.
+//
+// Zero ID is all zeros.
+//
+// Encode panics if dst length is less than 32.
```
Comment (cthulhu-rider) [inline container/id/id.go:41]: In current implementation it doesn't.

File: container/id/id.go [hunk 9]
```diff
@@ -1,90 +1,112 @@
...
-// Returns error if s is not a base58 encoded
-// ID data.
-func (id *ID) Parse(s string) error {
-	data, err := base58.Decode(s)
-	if err != nil {
-		return err
-	} else if len(data) != sha256.Size {
-		return errors.New("incorrect format of the string container ID")
+// Decode expects that src has 32 bytes length. If the input is malformed,
+// Decode returns an error. In this case src remains unchanged.
```
Comment (cthulhu-rider) [inline container/id/id.go:51]:
```suggestion
// Decode returns an error. In this case ID remains unchanged.
```
Statement that `src remains unchanged` is also needed.

File: container/id/id.go [hunk 10]
```diff
@@ -1,90 +1,112 @@
...
+func (id ID) String() string {
+	return id.EncodeToString()
 }
 
-// UnmarshalJSON decodes ID from protobuf JSON format.
-func (id *ID) UnmarshalJSON(data []byte) error {
-	return (*refs.ContainerID)(id).UnmarshalJSON(data)
+// Empty returns true if it is called on
+// zero container ID.
+func (id ID) Empty() bool {
```
Comment (cthulhu-rider) [inline container/id/id.go:110]: Why do we need it?

File: client/object_get.go (hunk 5 above)
Comment (carpawell) [reply client/object_get.go:38]: when i first time refactored (fixed after refactor) that code I tried to convert it to a container from v2, but then I realized that it is not a container

so at least we have one case of misunderstanding

Comment (cthulhu-rider) [reply client/object_get.go:38]: Ok, nvm, hope u'll do it more carefully next time.

File: container/id/id.go (hunk 8 above)
Comment (cthulhu-rider) [reply container/id/id.go]: On the contrary, I suggest leaving the panic, it's just that the implementation was inappropriate.

File: audit/result.go [hunk 11]
```diff
@@ -50,6 +56,39 @@ func (r *Result) Unmarshal(data []byte) error {
...
+
+	var (
+		oID   oid.ID
+		oidV2 refs.ObjectID
+	)
+
+	for _, oidV2 = range r.v2.GetPassSG() {
+		err = oID.ReadFromV2(oidV2)
+		if err != nil {
+			return fmt.Errorf("could not convert passed V2 container ID: %w", err)
```
Comment (cthulhu-rider) [inline audit/result.go:81]:
```suggestion
	return fmt.Errorf("invalid passed storage group ID: %w", err)
```

**Review by cthulhu-rider — APPROVED** [review]

**Review by cthulhu-rider — APPROVED** [review]

**Pull request merged**